data_structures/src/chain.rs

```diff
@@ -2559,15 +2586,48 @@ impl AltKeys {
     pub fn insert_keys_from_block(&mut self, block: &Block) {
         // Add miner bn256 public keys
         if let Some(value) = block.block_header.bn256_public_key.clone() {
-            self.insert_bn256(block.block_header.proof.proof.pkh(), value);
+            // If the provided key is valid, insert it in altkeys
+            if value.is_valid() {
+                self.insert_bn256(block.block_header.proof.proof.pkh(), value);
+            } else {
+                log::warn!(
+                    "Ignoring invalid bn256 public key {:02x?} inserted by PKH {:02x?} in the block",
+                    value, block.block_header.proof.proof.pkh()
+                 );
+            }
         }
         // Add bn256 public keys from commitment transactions
         for commit in &block.txns.commit_txns {
             if let Some(value) = commit.body.bn256_public_key.clone() {
-                self.insert_bn256(commit.body.proof.proof.pkh(), value);
+                // If the provided key is valid, insert it in altkeys
+                if value.is_valid() {
+                    self.insert_bn256(commit.body.proof.proof.pkh(), value);
+                } else {
+                    log::warn!(
+                        "Ignoring invalid bn256 public key {:02x?} inserted by PKH {:02x?} in the commit",
+                        value, block.block_header.proof.proof.pkh()
+                    );
+                }
             }
         }
     }
+
+    pub fn get_rep_ordered_bn256_list(
+        &self,
+        reputation_set: &TotalReputationSet<PublicKeyHash, Reputation, Alpha>,
+    ) -> Vec<Bn256PublicKey> {
+        let mut pkhs: Vec<PublicKeyHash> = self.bn256.iter().map(|(pkh, _)| *pkh).collect();
```

Comment:
why you sort 2 times?

Reply:
So that all nodes end up with the same reputation based sorting. We first sort by pkh, get the reputation of each node, then sort by reputation. In the case of two members having the same reputation, these will be sorted by pkh

Reply:
if you use sorted_by instead of sorted_by_key you has to use a Ordering ( the result of a cmp function). And if you do that, you can concat several comparitions using ".then()". And with that you avoid to sort 2 times.

![image](https://user-images.githubusercontent.com/44392885/84399979-c3508400-ac01-11ea-9e82-dcd3e4594073.png)

Reply:
You can use `sorted_by_key(|x| (x.a, x.b))` to sort by `x.a`, and in case of tie sort by `x.b`.

Since we want to sort by reputation and in case of tie by pkh, it should be `(reputation, pkh)`. But in any case it's easier to just add tests and see.

Reply:
I do not follow you guys

Reply:
` let ordered_alts: Vec<_> = pkhs
            .into_iter()
            .sorted_by_key(|&pkh| (reputation_set.get(&pkh), pkh))
            .filter_map(|pkh| self.get_bn256(&pkh).cloned())
            .collect();
`
Would this be sufficient then? Sort by rep, if tie sort by pkh